Handle extended thinking with Anthropic provider, process streaming thinking blocks and show them in the output tab

ferenci84 opened the pull request:
## Description

This modification allows users to see the human-readable parts of the thinking blocks in the LLM output.

This simple change is a first step on the way of full implementation of handling thinking blocks.

## Checklist

- [X] The relevant docs, if any, have been updated or created
- [X] The relevant tests, if any, have been updated or created

## Screenshots

[ For visual changes, include screenshots. ]
<img width="1291" alt="Screenshot 2025-03-01 at 14 18 00" src="https://github.com/user-attachments/assets/ceadec6a-37e1-4362-9694-0a2c49aeaaca" />

## Testing instructions

Thinking can be enabled in config this way:
```
   {
      "title": "Anthropic - Claude 3.7 Sonnet (thinking)",
      "model": "claude-3-7-sonnet-latest",
      "provider": "anthropic",
      "apiKey": "xxx",
      "completionOptions": {
        "temperature": 1,
        "topP": 0.999,
        "maxTokens": 8096
      },
      "requestOptions": {
        "extraBodyProperties": {
          "thinking": {
            "type": "enabled",
            "budget_tokens": 2048
          }
        }
      },
      "cacheBehavior": {
        "cacheConversation": true,
        "cacheSystemMessage": true
      }
    },
 ```

 Watch the Continue - LLM Prompt/Completion output tab.
 
 ## Next steps
 
1. The thinking mode doesn't currently work with tool use. The software should add the `thinking` and `redacted_thinking` blocks to the conversation history when sending subsequent requests. Tool use requires this, with no tool use, any thinking blocks will be ignored.

2. human-readable thinking blocks could be added to the UI in form of collapsible boxes. This is an example from TypingMind:
<img width="454" alt="Screenshot 2025-03-01 at 14 29 38" src="https://github.com/user-attachments/assets/5c2c82b2-c0fd-4d9d-9726-a51d098f4e11" />

## Edit

Handling thinking blocks and redacted_thinking blocks have been added. Due to these changes, these thinking blocks have also became visible. This may require some visual update, but it's now accessible and functional, and we are closer to a final solution where users can read the thinking blocks if they want.

## Additional testing instructions

The changes are working if you can call tools with no problem.

Redacted thinking can be tested by adding this string in your first prompt:
`ANTHROPIC_MAGIC_STRING_TRIGGER_REDACTED_THINKING_46C9A13E193C177646C7398A98432ECCCE4C1253D5E2D82641AC0E52CC2876CB`

## Edit 2

I have added rendering of thinking blocks as shown in the screenshots below:

<img width="491" alt="Screenshot 2025-03-05 at 11 42 56" src="https://github.com/user-attachments/assets/aafc6b34-6703-49bf-88bc-fd986714d936" />

<img width="494" alt="Screenshot 2025-03-05 at 11 43 05" src="https://github.com/user-attachments/assets/3327aa53-1edd-46e4-a1c7-d23fe31765cb" />

<img width="484" alt="Screenshot 2025-03-05 at 11 42 33" src="https://github.com/user-attachments/assets/5b091b54-5969-49c5-a6e5-310eeb3b91ad" />

chezsmithy commented:
Would it be possible to add the same support to Bedrock.ts now that I have enabled tools support, and have a seprate PR to enabled claude 3.7 on bedrock as well?

Symfomany commented:
Deep Research would be a killer-feature in this coding assistant Continue, it's possible to have Perplexity API Key too ?

ferenci84 commented:
@RomneyDa You are the assignee of issue https://github.com/continuedev/continue/issues/4339 and this PR would resolve it, can you please review?

SivasankarMathiyazhagan commented:
@ferenci84 - is there any calculations behind to set the budget tokens as **"budget_tokens": 2048** ? Could you please suggest how we can determine the appropriate budget token limit for our use case?

ferenci84 commented:
> @ferenci84 - is there any calculations behind to set the budget tokens as **"budget_tokens": 2048** ? Could you please suggest how we can determine the appropriate budget token limit for our use case?

You need to make sure that contextLength is larger than maxTokens and maxTokens is larger than budgetTokens, and budgetTokens is at least 1024. This uses 3200 as budget tokens: https://bigcode-bench.github.io/

sestinj commented:
@ferenci84 this looks really good, type updates and all. Two ways I think it could be improved:
- In this PR (https://github.com/continuedev/continue/pull/3840) we have a separate UI for displaying thinking tokens. I think we should keep the same UI for both for consistency
- as shown in this screenshot, the toolbar for [trash, copy, thumbs up, thumbs down] shows up between the input and thinking toggle, which causes some awkward blank space

<img width="754" alt="Screenshot 2025-03-09 at 8 38 43 PM" src="https://github.com/user-attachments/assets/1f058e56-4b67-4e43-b068-7cf26ba40813" />

Both of these thoughts lead to a bigger question of how we should combine the previous strategy for handling thinking tokens in Ollama (deepseek literally just outputs <thinking> tags in its output) with this one. It seems to me that parsing the <thinking> tags in `streamChat` in `llm/llms/index.ts` would be the best option

ferenci84 commented:
Yes, the big difference between claude and deepseek is that with anthropic the thinking and is a separate message (even multiple messages in case of redacted thinking) and this should be fed back to the AI in the same way on the next turn (which is required for tool use, otherwise those blocks are just ignored). I was thinking about combining the srategy, but I thought that it would be overly complicated due to the big differences. As of UI change it would be one more week until I could update, but why not let other users start sooner as it is right now?

ferenci84 commented:
The buttons between the first message and thinking block seems to be a mistake as it's not present in other cases.

ferenci84 commented:
Thinking further, there is one corner case in which a shared implementation would be helpful: let's say you start the conversation with anthropic and then swith to deepseek. Not sure how common is that.

ferenci84 commented:
My ide is to have a metadata object with dynamic data structure for assistant messages,  in which thinking data could be added, but also other possibly complex data that help us reconstruct the messages. This could even be any type so that any additional data like signature, multiple redacted thinking that shouldn't be displayed, could be stored. In addition we could add an optional "display" key that would contain data about how to display, for example that deepseek will not show think tag in the main message, we could have sepaeate (optional) display information for main message and thinking message, that is parsed in the core (even better, in the LLM class as it would better help encapsulation), not in the UI as is is now. What do you think?

Any way, I think crucial updates should be done (for example those that support seeing thinking blocks and avoid errors with tools) because it would help the product to succeed.

FallDownTheSystem commented:
I didn't realize someone else was already working on this. 😅

I've basically implemented the same functionality, with tool support and using the existing Reasoning UI: https://github.com/continuedev/continue/pull/4559

Although there's no support for Bedrock in mine. Also I tried and you can swap between thinking models, unless you use tools, then when you try to swap to DeepSeek you get an error.

Anyway, I'm happy with either or any solution being merged into Continue. 😄

ferenci84 commented:
@FallDownTheSystem Checked your solution, I also wanted to go the same path of adding a new type of MessagePart, before I realized that it would require change in lot of seemingly irrelevant code. What do you think about a longterm solution to simplify as I explained above? Not sure if it was clear. For short term, I think just one of the working solution should be merged so that ordinary users can start to use the new feature of Claude, they will not care as long as it's working well.

FallDownTheSystem commented:
> @FallDownTheSystem Checked your solution, I also wanted to go the same path of adding a new type of MessagePart, before I realized that it would require change in lot of seemingly irrelevant code. What do you think about a longterm solution to simplify as I explained above? Not sure if it was clear. For short term, I think just one of the working solution should be merged so that ordinary users can start to use the new feature of Claude, they will not care as long as it's working well.

It does feel like some kind of refactoring would be in order. Currently the message are converted into suitable format for the UI and then converted back for the API. Ideally you'd be able to keep the original data intact and have LLM specific methods that are responsible for converting that to a format that's suitable for the UI.

I can see that the code base has gotten more complex as support for new formats have been added. I'm sure originally everything was pretty simple when the output was just a string.

ferenci84 commented:
@sestinj I have changed it to a button display and also added seconds:
<img width="353" alt="Screenshot 2025-03-11 at 22 19 51" src="https://github.com/user-attachments/assets/394d5c65-dc1a-4031-bf06-3d7281db4111" />

It looks like this when seconds is not available:
![Uploading Screenshot 2025-03-11 at 22.21.03.png…]()

When opened:
<img width="357" alt="Screenshot 2025-03-11 at 22 23 22" src="https://github.com/user-attachments/assets/d3ae6685-d882-4dbd-a273-a20a2fb34a50" />

It's also part of the PR with Bedrock support: https://github.com/continuedev/continue/pull/4453
